### PR TITLE
Add code fix to remove unreachable code

### PR DESCRIPTION
**Pull request by andy-ms**
Mentioned in #24026

### CODE REVIEW

File: src/services/codefixes/fixUnreachableCode.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+    function getStatement(sourceFile: SourceFile, start: number): Statement {
+        const token = getTokenAtPosition(sourceFile, start, /*includeJsDocComment*/ false);
+        const statement = findAncestor(token, isStatement);
+        Debug.assert(statement.getStart(sourceFile) === token.getStart(sourceFile));
+
+        const container = (isBlock(statement.parent) ? statement.parent : statement).parent;
+        switch (container.kind) {
+            case SyntaxKind.IfStatement:
+                return (container as IfStatement).elseStatement ? statement : container as IfStatement;
```
Comment (mhegazy) [inline src/services/codefixes/fixUnreachableCode.ts]: if the statement was not in a block, that will leave the the ifStatement syntactically incorrect.. e.g.
```ts
    if (false) console.log();
    else console.log();
```

File: src/services/codefixes/fixUnreachableCode.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+            case SyntaxKind.WhileStatement:
+            case SyntaxKind.ForStatement:
+                return container as WhileStatement | ForStatement;
+            default:
+                return statement;
```
Comment (mhegazy) [inline src/services/codefixes/fixUnreachableCode.ts]: you want to return all statemetns from this one untill the end of the block, e.g.:

```ts
function f() { 
   return 0;

  // both should be removed, error is only on the first.
   var x = 0;
   console.log(x);
}
```

File: src/services/codefixes/fixUnreachableCode.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                    break;
+                }
+                // falls through
+            case SyntaxKind.WhileStatement:
+            case SyntaxKind.ForStatement:
+                changes.deleteNode(sourceFile, container);
+                break;
+            default:
+                if (isBlock(statement.parent)) {
+                    split(sliceAfter(statement.parent.statements, statement), s => !isFunctionDeclaration(s), (start, end) => changes.deleteNodeRange(sourceFile, start, end));
```
Comment (ajafff) [inline src/services/codefixes/fixUnreachableCode.ts]: You probably also want to preserve type alias declarations, interfaces, const enums, type only namespaces, var statements, ...

File: src/services/codefixes/fixUnreachableCode.ts [hunk 4]
```diff
@@ -36,14 +36,33 @@ namespace ts.codefix {
...
                     changes.deleteNode(sourceFile, statement);
                 }
         }
     }
 
+    function shouldRemove(s: Statement): boolean {
+        // Don't remove statements that can validly be used before they appear.
+        return !isFunctionDeclaration(s) && !isPurelyTypeDeclaration(s) &&
+            // `var x;` may declare a variable used above
+            !(isVariableStatement(s) && !(getCombinedNodeFlags(s) & (NodeFlags.Let | NodeFlags.Const)) && s.declarationList.declarations.some(d => !d.initializer));
```
Comment (mhegazy) [inline src/services/codefixes/fixUnreachableCode.ts]: if it is a var statement it can not have an initializer, and it is not ambient.

Comment (andy-ms) [reply src/services/codefixes/fixUnreachableCode.ts]: If it's ambient, it should not have an initializer so we won't remove it, right?

File: tests/cases/fourslash/codeFixUnreachableCode_if.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+////    a;
+////}
+
+verify.codeFixAll({
+    fixId: "fixUnreachableCode",
+    fixAllDescription: "Remove all unreachable code",
+    newFileContent:
+`
+// No good way to delete just the 'if' part
+if (false) { } else b;
```
Comment (amcasey) [inline tests/cases/fourslash/codeFixUnreachableCode_if.ts]: In a perfect world, the output would be `b;`?

File: src/services/codefixes/fixUnreachableCode.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        const token = getTokenAtPosition(sourceFile, start, /*includeJsDocComment*/ false);
+        const statement = findAncestor(token, isStatement);
+        Debug.assert(statement.getStart(sourceFile) === token.getStart(sourceFile));
+
+        const container = (isBlock(statement.parent) ? statement.parent : statement).parent;
+        switch (container.kind) {
+            case SyntaxKind.IfStatement:
+                if ((container as IfStatement).elseStatement) {
+                    if (isBlock(statement.parent)) {
+                        changes.deleteNodeRange(sourceFile, first(statement.parent.statements), last(statement.parent.statements));
```
Comment (amcasey) [inline src/services/codefixes/fixUnreachableCode.ts]: Don't we want to eliminate the whole else-clause?

File: src/services/codefixes/fixUnreachableCode.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+                    split(sliceAfter(statement.parent.statements, statement), shouldRemove, (start, end) => changes.deleteNodeRange(sourceFile, start, end));
+                }
+                else {
+                    changes.deleteNode(sourceFile, statement);
+                }
+        }
+    }
+
+    function shouldRemove(s: Statement): boolean {
+        // Don't remove statements that can validly be used before they appear.
```
Comment (amcasey) [inline src/services/codefixes/fixUnreachableCode.ts]: I'm missing something - why would it be reported as unused if it's used above?

File: src/services/codefixes/fixUnreachableCode.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+            case SyntaxKind.WhileStatement:
+            case SyntaxKind.ForStatement:
+                changes.deleteNode(sourceFile, container);
+                break;
+            default:
+                if (isBlock(statement.parent)) {
+                    split(sliceAfter(statement.parent.statements, statement), shouldRemove, (start, end) => changes.deleteNodeRange(sourceFile, start, end));
+                }
+                else {
+                    changes.deleteNode(sourceFile, statement);
```
Comment (amcasey) [inline src/services/codefixes/fixUnreachableCode.ts]: Will a switch case ever be reported as unused?  Does that need special handling?  What about a catch block?

**Pull request merged**